### PR TITLE
Migrate to Typescript(#14) but better

### PR DESCRIPTION
**Pull request by JcdeA**
#14 , but better.

### CODE REVIEW

**Comment by JcdeA** [comment]
@LiveDuo I've separated the Tailwind stuff, I'll create a new pr for that.

**Comment by JcdeA** [comment]
@LiveDuo Please let me know If ypu have anything you want me to change

File: .gitignore
Comment (LiveDuo) [inline .gitignore:17]: Although this is not an issue, we want to replicate a real environment with the `nextjs-project` so it's better to leave the `default.json file` in.

PS: If this becomes annoying with local development we can add update husky config to ignore it.

File: .gitignore
Comment (LiveDuo) [inline .gitignore:37]: Can we move this to `lib/destack/.gitignore`?

File: .tsbuildinfo
Comment (LiveDuo) [inline .tsbuildinfo:1]: Not sure why this stayed here since it's added to `.gitignore`.

Maybe this would help: [https://stackoverflow.com/questions/1274057/how-to-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore](https://stackoverflow.com/questions/1274057/how-to-make-git-forget-about-a-file-that-was-tracked-but-is-now-in-gitignore)

File: .vscode/settings.json
Comment (LiveDuo) [inline .vscode/settings.json:1]: Same as above!

File: dev/nextjs-project/data/default.json
Comment (LiveDuo) [inline dev/nextjs-project/data/default.json:1]: I think this file should be deleted.

In VSCode it's there an option to quickly discard changes you rather not commit.

image

File: lib/destack/.tsbuildinfo
Comment (LiveDuo) [inline lib/destack/.tsbuildinfo:1]: Should be remove too!

File: lib/destack/rollup.config.js
Comment (LiveDuo) [inline lib/destack/rollup.config.js:28]: Why this is changed?

Not a problem but wondering if it created any issues.

File: lib/destack/rollup.config.js
Comment (LiveDuo) [inline lib/destack/rollup.config.js:29]: I'm working on Github actions to automatically run tests with different node.js versions. I'll also manually test it with at least Node 10 right before merging the PR.

For now, I think the target be `node10` to make Destack available to previous versions of Node and Next.js.

File: lib/destack/rollup.config.js
Comment (LiveDuo) [inline lib/destack/rollup.config.js:66]: > I'm working on Github actions to automatically run tests with different node.js versions. I'll also manually test it with at least Node 10 right before merging the PR.
> 
> For now, I think the target be `node10` to make Destack available to previous versions of Node and Next.js.

**Review by LiveDuo — CHANGES_REQUESTED** [review]
We will need a few changes but nothing major. It's a lot of small changes and cleaning up committed files that are now in `.gitignore`.

Nice work! Can't wait to get TS merged 🛠

File: dev/nextjs-project/data/default.json
Comment (JcdeA) [reply dev/nextjs-project/data/default.json:1]: I'm trying to resolve this, but wouldn't `git rm --cached` delete the file from remote too? I'm looking into how I might undo the diffs..

Comment (LiveDuo) [reply dev/nextjs-project/data/default.json:1]: I think so. Do you see any side effect of removing the file?

File: lib/destack/rollup.config.js
Comment (LiveDuo) [reply lib/destack/rollup.config.js]: That's very strange.

Did a quick check with esbuild on Github and I can't find anywhere to explicitly check for specific node versions. I guess that means parse the version from the string eg. `nodex.y.z`

Comment (LiveDuo) [reply lib/destack/rollup.config.js]: Turns out `node10` is the default target for esbuild so it should be fine [esbuild/scripts/esbuild.js](https://github.com/evanw/esbuild/blob/3431cc835b6dcf8f6172665a68b1d9c8b80fb147/scripts/esbuild.js#L11).

File: .gitignore
Comment (LiveDuo) [reply .gitignore]: `lib/destack` just has its own .gitignore to make things a bit clearer. There isn't any special configuration to it, it just works!

File: dev/nextjs-project/data/default.json
Comment (JcdeA) [reply dev/nextjs-project/data/default.json:1]: The user would have to create the same page to pass build.test.js.

File: lib/destack/rollup.config.js
Comment (JcdeA) [reply lib/destack/rollup.config.js]: Great!

File: lib/destack/.gitignore
Comment (LiveDuo) [inline lib/destack/.gitignore:36]:
```suggestion
.tsbuildinfo
```

**Review by LiveDuo — APPROVED** [review]
Time to get type-checking!

Will do a last manual check when I get back and merge right after 🚀

File: lib/destack/.gitignore
Comment (LiveDuo) [reply lib/destack/.gitignore:36]: @JcdeA Don't worry about that little thing, will sort it out later.

Thanks for bringing typescript to Destack!

**Comment by JcdeA** [comment]
On huge problem is that grapesjs isn't fully typed yet..

There used to be a community effort but it seems to be abandoned.

**Comment by LiveDuo** [comment]
> On huge problem is that grapesjs isn't fully typed yet..
> 
> There used to be a community effort but it seems to be abandoned.

That's very true and I was thinking about that a bit the last few days.

A migration that seems to has a lot of merits for the project is using [craft.js](https://github.com/prevwong/craft.js) instead of Grapes.js.

Craft.js is a page builder library similar to Grapes.js that is built with React and Typescript and seems like the perfect match. It won't be too much hassle because Craft is inspired by Grapes and works similarly in many respects.

**Comment by JcdeA** [comment]
Thanks for merging the pr, this is one of my first contributions to a relatively big project!

> Craft.js is a page builder library similar to Grapes.js that is built with React and Typescript and looks like the perfect match. It won't be too much hassle because Craft is inspired by Grapes and works similarly in many respects.
That's a great Idea! I'll try migrating to it when I have the time. 
But for now, I have finals in 3 weeks (I'm 14yo), so I won't have much time to develop things for a while.

**Comment by LiveDuo** [comment]
> Thanks for merging the pr, this is one of my first contributions to a relatively big project!
> 
> > Craft.js is a page builder library similar to Grapes.js that is built with React and Typescript and looks like the perfect match. It won't be too much hassle because Craft is inspired by Grapes and works similarly in many respects.
> > That's a great Idea! I'll try migrating to it when I have the time.
> > But for now, I have finals in 3 weeks (I'm 14yo), so I won't have much time to develop things for a while.

Glad that it happened with Destack. Everything worked flawlessly!

I will move to adding a colouring scheme for now similar to https://tailblocks.cc/. That will make sites build with Destack more custom and closer to user's taste. It also brings back the conversion about custom tailwind config and customisation in general with Destack. 

If it's an easy thing to do it will be helpful to have the custom tailwind config PR. I'll try to consider different solutions and add a few thoughts on the PR to discuss further.

Good luck with your finals!

**Pull request merged**
